### DISCUSSION TITLE
[Deployment] Deploy contract onto zkSync with Viem

### DISCUSSION THREAD

**Discussion by spatializes** [zkSync-Community-Hub/zksync-developers#425] [Deployments]
### Environment

Mainnet

### zkSolc Version

n/a

### zksync-web3 Version

n/a

### Hardhat.config.ts

```javascript
n/a
```

### Deployment Script (WITHOUT PRIVATE KEY)

```javascript
import { createWalletClient, custom } from 'viem'
import { zkSync } from 'viem/chains'

 
const walletClient = createWalletClient({ 
  chain: zkSync, 
  transport: custom(window.ethereum!), 
})

const hash = await walletClient.deployContract({
    abi: abi,
    args: args,
    bytecode: bytecode,
    ...options,
})
```

### Package.json

```json
n/a
```

### Contract Code

```solidity

// SPDX-License-Identifier: MIT
pragma solidity ^0.8.20;

import "@openzeppelin/contracts/token/ERC20/ERC20.sol";
import "@openzeppelin/contracts/token/ERC20/extensions/ERC20Burnable.sol";
import "@openzeppelin/contracts/token/ERC20/extensions/ERC20Pausable.sol";
import "@openzeppelin/contracts/access/Ownable.sol";
import "@openzeppelin/contracts/token/ERC20/extensions/ERC20Permit.sol";

contract BRTST is ERC20, ERC20Burnable, ERC20Pausable, Ownable, ERC20Permit {
    constructor(address initialOwner)
        ERC20("BRTSTZKSYNC", "MBRTSTTK")
        Ownable(initialOwner)
        ERC20Permit("BRTSTZKSYNC")
    {}

    function pause() public onlyOwner {
        _pause();
    }

    function unpause() public onlyOwner {
        _unpause();
    }

    function mint(address to, uint256 amount) public onlyOwner {
        _mint(to, amount);
    }

    // The following functions are overrides required by Solidity.

    function _update(address from, address to, uint256 value)
        internal
        override(ERC20, ERC20Pausable)
    {
        super._update(from, to, value);
    }
}
```

### Does this work on other EVMs? (If yes, please list at least 1 of them)

Yes, Ethereum mainnet, Arbitrum One, Optimism, Avalanche, Polygon, Metis, Celo, Inco, fEVM, etc

### Description of What Your Contract Does

Just an ERC20 example

### Additional Details

Our dapp is currently running on 10+ evms, where users can deploy contacts through the dapp. We're using viem and wagmi, with the code above. Does zksync support deploys using viem in the same way? Is there something else I should be doing? I'm getting crazy gas estimates ($300) and failures (Internal JSON-RPC error) even if we try to deploy with the high estimate.

**Comment by AlbionaHoti** [comment]
Hi @spatializes, yes, you can use viem to interact seamlessly with smart contracts deployed on zkSync. Have you seen this [guide](https://docs.zksync.io/build/tutorials/tooling-guides/viem.html#viem)?

**Reply by spatializes** [reply]
Hey @AlbionaHoti! I did see the guide, as well as this guide directly with viem (https://viem.sh/zksync). However, I am not able to test interaction because I'm not able to deploy. It looks like these guides both walk through interactions, but do not show how to deploy a smart contract with zkSync using viem.

**Reply by dutterbutter** [reply]
Hey @spatializes have you reviewed the `deployContract` action for zkSync in viem? 👉  https://viem.sh/zksync/actions/deployContract 

We just had this merged, so please let us know if you run into an issue or have suggested improvements. We will be providing more verbose guides in the near future!

**Reply by dutterbutter** [reply]
@spatializes can you try increasing the value of `gasPerPubdata` to something much higher just to test. Also ensure the fields `bytecode` and `value` are correctly passing the values expected.

**Reply by dutterbutter** [reply]
Can you share the repo by any chance?

**Reply by dutterbutter** [reply]
@Solda1219 thank you. 

I cant reproduce the issue easily from the provided gist but one thing its important to note, is that zkSync bytecode is different then other chains. The bytecode expected when using viem `deployContract` actions is the bytecode output when compiled using our custom compiler `zksolc`. You can make use of our [hardhat-zksync-solc plugin](https://docs.zksync.io/build/tooling/hardhat/hardhat-zksync-solc.html) to compile and ensure the correct bytecode is used. To be precise, you cannot reuse the same bytecode from evm compilation (other chains) as we make use of our zkevm. We intend to continue to add to our upstream viem support with utility methods to make this easier to obtain as its currently a work-in-progress. We apologize for any inconvenience here.  

I don't have a repo readily available for viem deployContract action but I can provide one next week. In the meantime, here are the [tests](https://github.com/wevm/viem/blob/main/src/zksync/actions/deployContract.test.ts) ensuring the action works as intended.

**Comment by dutterbutter** [comment]
Duplicate of https://github.com/zkSync-Community-Hub/zksync-developers/discussions/433

**Reply by itxtoledo** [reply]
they deleted this issue, how can I fix that?
